src/vsr/replica.zig

```diff
@@ -783,65 +837,69 @@ pub fn ReplicaType(
-        //
-        // In the second case we avoid setting the view to make sure pings can still be sent
-        // during view changes.
+        /// Pings are used by replicas to synchronise cluster time and to probe for network connectivity.
         fn on_ping(self: *Self, message: *const Message) void {
+            assert(message.header.command == .ping);
             if (self.status != .normal and self.status != .view_change) return;
 
-            assert(self.status == .normal or self.status == .view_change);
+            if (message.header.replica == self.replica) {
+                log.warn("{}: on_ping: ignoring (self)", .{self.replica});
+                return;
+            }
 
             // TODO Drop pings that were not addressed to us.
 
-            var pong = Header{
+            self.send_header_to_replica(message.header.replica, .{
                 .command = .pong,
                 .cluster = self.cluster,
                 .replica = self.replica,
-            };
-
-            if (message.header.client > 0) {
-                assert(message.header.replica == 0);
-
-                // We must only ever send our view number to a client via a pong message if we are
-                // in normal status. Otherwise, we may be partitioned from the cluster with a newer
-                // view number, leak this to the client, which would then pass this to the cluster
-                // in subsequent client requests, which would then ignore these client requests with
-                // a newer view number, locking out the client. The principle here is that we must
-                // never send view numbers for views that have not yet started.
-                if (self.status == .normal) {
-                    pong.view = self.view;
-                    self.send_header_to_client(message.header.client, pong);
-                }
-            } else {
-                assert(message.header.view == 0);
-
-                if (message.header.replica == self.replica) {
-                    log.warn("{}: on_ping: ignoring (self)", .{self.replica});
-                } else {
-                    // Copy the ping's monotonic timestamp to our pong and add our wall clock sample:
-                    pong.op = message.header.op;
-                    pong.timestamp = @bitCast(u64, self.clock.realtime());
-                    self.send_header_to_replica(message.header.replica, pong);
-                }
-            }
+                // Copy the ping's monotonic timestamp to our pong and add our wall clock sample:
+                .op = message.header.op,
+                .timestamp = @bitCast(u64, self.clock.realtime()),
+            });
         }
 
         fn on_pong(self: *Self, message: *const Message) void {
-            if (message.header.client > 0) return;
+            assert(message.header.command == .pong);
             if (message.header.replica == self.replica) return;
```

Comment:
Hm, reading `on_pong` makes me appreciate ` // TODO Drop pings that were not addressed to us.` comment more. It seems that a misdirected ping/pong could make our clock go askew. I guess we want to stuff target replica into `.context`?

Reply:
if the `pong` is misdirected, than original timestamp is not what _we_ sent, but rather what some other replica sent.

Reply:
Ooh, I see